### PR TITLE
MON-4046: add early validation for Platform and UWM monitoring configmaps

### PR DESCRIPTION
**Pull request by machine424**
This is done through validation webhooks.

An opt-out mechansim is provided to make the configmaps skip CMO's
validate webhook in case.

The exposed validate webhook path has no auth for now.

* [ ] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -100,6 +103,13 @@ func (s *Server) Run(ctx context.Context) error {
 		return err
 	}
 
+	// This is a temporary measure until the CRD-based configuration is GA.
+	// Following tge fail-early paradigm makes configuration failures easily detectable by users.
+	// This will also aid in the transition to CRD by providing a preview of the future configuration process.
+	server.Handler.NonGoRestfulMux.HandleFunc("/admission-monitoringconfigmaps/validate", func(w http.ResponseWriter, r *http.Request) {
```
Comment (machine424) [inline pkg/server/server.go:109]: https://github.com/openshift/cluster-monitoring-operator/blob/8319fc2d5eaa36929e28e261337a78e7dab39920/manifests/0000_50_cluster-monitoring-operator_03-service.yaml#L10

File: pkg/configvalidate/configvalidate.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+			Message: err.Error(),
+		},
+	}
+}
+
+type parseConfig func(s string) error
+
+func configParser() parseConfig {
+	return func(s string) error {
+		_, err := manifests.NewUserConfigFromString(s)
```
Comment (machine424) [inline pkg/configvalidate/configvalidate.go:109]: check if it's strict: duplicated fields? unknwon fields? etc

File: pkg/manifests/config.go [hunk 3]
```diff
@@ -189,13 +189,16 @@ func (cps CollectionProfiles) String() string {
 	return sb.String()
 }
 
-func NewConfig(content io.Reader, collectionProfilesFeatureGateEnabled bool) (*Config, error) {
+func NewConfig(content []byte, collectionProfilesFeatureGateEnabled bool) (*Config, error) {
 	c := Config{CollectionProfilesFeatureGateEnabled: collectionProfilesFeatureGateEnabled}
 	cmc := defaultClusterMonitoringConfiguration()
-	err := k8syaml.NewYAMLOrJSONDecoder(content, 4096).Decode(&cmc)
+	// TODO: say unknwon not fallged currenttly.
```
Comment (machine424) [inline pkg/manifests/config.go:195]: release notes

File: pkg/server/server.go [hunk 4]
```diff
@@ -90,6 +97,9 @@ func (s *Server) Run(ctx context.Context) error {
 		// prefix the authorizer with the permissions for metrics scraping which are well known.
 		// openshift RBAC policy will always allow this user to read metrics.
 		hardcodedauthorizer.NewHardCodedMetricsAuthorizer(),
+		// disable authz on the validate webhook paths.
+		// TODO: explain why
```
Comment (machine424) [inline pkg/server/server.go:101]: todo

File: test/e2e/config_test.go [hunk 5]
```diff
@@ -57,6 +57,11 @@ func TestClusterMonitoringOperatorConfiguration(t *testing.T) {
 			"config.yaml": `cannot be deserialized`,
 		},
 	}
+	// TODO: maybe check error type
```
Comment (machine424) [inline test/e2e/config_test.go:60]: todo

File: pkg/configvalidate/configvalidate.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+func uwmConfigParser() parseConfig {
+	return func(c *corev1.ConfigMap) error {
+		_, err := manifests.NewUserConfigFromConfigMap(c)
+		return err
+	}
+}
+
+type configmapsValidator struct {
+	// TODO: make sure collectionProfilesEnabled doesn't change
```
Comment (machine424) [inline pkg/configvalidate/configvalidate.go:56]: todo

File: test/e2e/user_workload_monitoring_test.go [hunk 7]
```diff
@@ -46,6 +46,9 @@ type scenario struct {
 	assertion func(*testing.T)
 }
 
+// TODO: adapt
```
Comment (machine424) [inline test/e2e/user_workload_monitoring_test.go:49]: todo

File: pkg/manifests/manifests_test.go [hunk 8]
```diff
@@ -650,9 +650,6 @@ func TestPrometheusOperatorConfiguration(t *testing.T) {
 	c, err := NewConfigFromString(`prometheusOperator:
   nodeSelector:
     type: master
-  image: quay.io/test/prometheus-operator
```
Comment (machine424) [inline pkg/manifests/manifests_test.go:653]: thanks to the new `UnmarshalStrict`

File: pkg/server/server.go [hunk 9]
```diff
@@ -101,6 +110,15 @@ func (s *Server) Run(ctx context.Context) error {
 		return err
 	}
 
+	// This is a temporary measure until the CRD-based configuration is GA.
+	// Following tge fail-early paradigm makes configuration failures easily detectable by users.
```
Comment (jan--f) [inline pkg/server/server.go:114]:
```suggestion
	// Following the fail-early paradigm this makes configuration failures easily detectable by users.
```

**Review by jan--f — CHANGES_REQUESTED** [review]
Just a nit to fix a comment typo. Otherwise this lgtm!

**Comment by jan--f** [comment]
/lgtm
I'll give @simonpasquier a chance to take a look too so
/hold
feel free to unhold

**Pull request merged**